platform-mbean/src/main/java/org/jboss/as/platform/mbean/AbstractExtendedMBean.java

```diff
@@ -0,0 +1,116 @@
+import javax.management.IntrospectionException;
+import javax.management.InvalidAttributeValueException;
+import javax.management.MBeanAttributeInfo;
+import javax.management.MBeanException;
+import javax.management.MBeanOperationInfo;
+import javax.management.MBeanParameterInfo;
+import javax.management.MalformedObjectNameException;
+import javax.management.ObjectName;
+import javax.management.ReflectionException;
+import org.jboss.as.controller.OperationFailedException;
+
+/**
+ * Utility to access to the com.sun.management MBeans
+ *
+ * @author jdenise
+ */
+class AbstractExtendedMBean {
+
+    private final ObjectName name;
+
+    AbstractExtendedMBean(String name) throws OperationFailedException {
+        try {
+            this.name = new ObjectName(name);
+        } catch (MalformedObjectNameException ex) {
+            throw new OperationFailedException(ex);
+        }
+    }
+
+    boolean isAttributeDefined(String attributeName) {
+        Objects.requireNonNull(attributeName);
+        try {
+            for (MBeanAttributeInfo attribute
+                    : ManagementFactory.getPlatformMBeanServer().getMBeanInfo(name).getAttributes()) {
+                if (attributeName.equals(attribute.getName())) {
+                    return true;
+                }
+            }
+        } catch (InstanceNotFoundException | IntrospectionException | ReflectionException ex) {
+            // XXX OK, attribute not available.
+        }
+        return false;
+    }
+
+    // Operation
+    boolean isOperationDefined(String operationName, String[] types) {
+        Objects.requireNonNull(operationName);
+        Objects.requireNonNull(types);
+        try {
+            for (MBeanOperationInfo op
+                    : ManagementFactory.getPlatformMBeanServer().getMBeanInfo(name).getOperations()) {
+                if (operationName.equals(op.getName())) {
+                    List<String> signature = new ArrayList<>();
+                    for(MBeanParameterInfo p : op.getSignature()) {
+                        signature.add(p.getType());
+                    }
+                    List<String> receivedTypes =  Arrays.asList(types);
+                    if(receivedTypes.equals(signature)) {
+                        return true;
+                    }
+                }
+            }
+        } catch (InstanceNotFoundException | IntrospectionException | ReflectionException ex) {
+            // XXX OK, attribute not available.
+        }
+        return false;
+    }
+
+    Object getAttribute(String attributeName) throws OperationFailedException {
+        try {
+            return ManagementFactory.getPlatformMBeanServer().
+                    getAttribute(name, attributeName);
+        } catch (AttributeNotFoundException
+                | InstanceNotFoundException
+                | MBeanException
+                | ReflectionException ex) {
+            throw new OperationFailedException(ex);
```

Comment:
This comment applies to this commit overall; I just randomly picked this line.

We should only throw OFE or any other exception type that implements OperationClientException if the failure indicates an end user mistake. For any other cases, an OperationStepHandler should throw a runtime exception of some sort.

The server kernel exception handling relies on this, i.e. perhaps dealing with user mistakes one way while handling other failures differently.